coreax/util.py

```diff
@@ -282,19 +308,100 @@ def jit_test(
         return fn(*args, **kwargs)
 
-    assert hash(_fn) != hash(fn), "Cannot guarantee recompilation of `fn`."
+    if check_hash:
+        assert hash(_fn) != hash(fn), "Cannot guarantee recompilation of `fn`."
 
     start_time = time.time()
     block_until_ready(_fn(*fn_args, **fn_kwargs))
     end_time = time.time()
     pre_delta = end_time - start_time
+
     start_time = time.time()
     block_until_ready(_fn(*fn_args, **fn_kwargs))
     end_time = time.time()
     post_delta = end_time - start_time
+
     return pre_delta, post_delta
 
 
+def _format_number(num: float) -> str:
+    """Standardise the format of the input number."""
+    if num == 0:
+        return "0 s"
+    order = log10(num)
+    if order >= 2:  # noqa: PLR2004
+        scaled_num = num / 60
+        order = "mins"
+    elif 0 <= order < 2:  # noqa: PLR2004
+        scaled_num = num
+        order = "s"
+    elif -3 <= order < 0:  # noqa: PLR2004
+        scaled_num = 1e3 * num
+        order = "ms"
+    elif -6 <= order < -3:  # noqa: PLR2004
+        scaled_num = 1e6 * num
+        order = "\u03bcs"
+    elif -9 <= order < -6:  # noqa: PLR2004
+        scaled_num = 1e9 * num
+        order = "ns"
+    else:
+        scaled_num = 1e12 * num
+        order = "ps"
+
+    return f"{round(scaled_num, 2)} {order}"
+
+
+def speed_comparison_test(
+    function_setups: Sequence[JITCompilableFunction],
+    num_runs: int = 10,
+    log_results: bool = False,
+    check_hash: bool = False,
+) -> tuple[list[tuple[Array, Array]], dict[str, float]]:
+    """
+    Compare compilation time and runtime of a list of JIT-able functions.
+
+    :param function_setups: Sequence of instances of :class:`JITCompilableFunction`
+    :param num_runs: Number of times to average function timings over
+    :log_results: If :data:`True`, the results are formatted and logged
+    :param check_hash: If :data:`True` check that the hash of the JITted functions are
+        different to the supplied functions.
+    :return: Mean and standard deviation of compilation time and runtime for each
+        function as a list of tuples, and a dictionary of raw timings
+    """
+    num_functions = len(function_setups)
+    timings_dict = {}
+    results = []
+    for i in range(num_functions):
+        timings = jnp.zeros((num_runs, 2))
+        for j in range(num_runs):
+            timings = timings.at[j, :].set(
+                jit_test(*function_setups[i], check_hash=check_hash)
+            )
+        # Compute the time just spent on compilation
+        timings_dict[f"timings_{i}"] = timings.at[:, 0].set(
+            timings[:, 0] - timings[:, 1]
+        )
+        # Compute summary statistics
+        results.append((timings.mean(axis=0), timings.std(axis=0)))
+
+    if log_results:
+        for k in range(num_functions):
+            _logger.info("------------------- Function %s -------------------", k + 1)
+            _logger.info(
+                "Compilation time: "
```

Comment:
`Virgin` or `Uncompiled`? `Compilation time` is really the time stored in `timings_dict`.

Reply:
My understanding of what JAX is doing when you run the function the first time is compiling the function, then dispatching it. Then the second run just dispatches the compiled code. So by taking the difference between the second and first times we get an estimate of compilation time?

Reply:
Let's talk.